storage/SMAIndexSubBlock.hpp

```diff
@@ -0,0 +1,432 @@
+   * and should be used.
+   *
+   * @param predicate A simple predicate too be evaluated.
+   * 
+   * @return The cost associated with the type of match. Empty matches are constant,
+   *         partial matches require a scan, and complete matches require something
+   *         less than a regular scan (because we don't need to do the comparison
+   *         operation for each tuple) but is still linear with the number of tuples.
+   */
+  predicate_cost_t estimatePredicateEvaluationCost(
+      const ComparisonPredicate &predicate) const override;
+
+  /**
+   * @warning Calling this method on the SMA index implies that we are not going
+   *          to do a scan for some tuple matches. As in, the SMA index will
+   *          either return an empty set of tuple ids, or a set of tuple ids
+   *          which is the entire set of all tuple ids in the storage subblock.
+   * @note Currently this version only supports simple comparisons of a literal
+   *       value with a non-composite key.
+   **/
+  TupleIdSequence* getMatchesForPredicate(const ComparisonPredicate &predicate,
+                                          const TupleIdSequence *filter) const override;
+
+  /**
+   * @brief Update the index to reflect the current state of the storage block.
+   *
+   * @return \c true if successful.
+   */
+  bool rebuild() override;
+
+  /**
+   * @brief Returns if the index is consistent. Rebuilding will ensure this
+   *        returns true.
+   *
+   * @return \c true if inconsistent (rebuild to return true).
+   */
+  bool requiresRebuild() const;
+
+  /**
+   * @brief Given an attribute, quickly check to see if the SMA index contains an
+   *        entry for it.
+   *
+   * @param attribute The ID of the attribute to check.
+   * 
+   * @return \c true if this index contains an entry.
+   */
+  bool hasEntryForAttribute(attribute_id attribute) const;
+
+  /**
+   * @brief Returns the number of tuples, the aggregate COUNT, of the storage sub block.
+   *
+   * @return Number of tuples in the sub block.
+   */
+  std::uint32_t getCount() const {
+    return reinterpret_cast<sma_internal::SMAHeader*>(sub_block_memory_)->count_aggregate;
+  }
+
+ private:
+  inline sma_internal::Selectivity selectivityForPredicate(const ComparisonPredicate &predicate) const;
+
+  // Retrieves an entry, first checking if the given attribute is indexed.
+  inline const sma_internal::SMAEntry* getEntryChecked(attribute_id attribute) const {
+    DCHECK(attribute > -1) << "Attribute Id must be >= 0";
+
+    if (total_attributes_ > static_cast<std::size_t>(attribute)) {
+      int index = attribute_to_entry_[attribute];
+      if (index != -1) {
+        return entries_ + index;
+      }
+    }
+    return nullptr;
+  }
+
+  // Retrieves an entry, not checking if the given attribute is indexed.
+  // Warning: This should not be used unless the attribute is indexed.
+  inline const sma_internal::SMAEntry* getEntry(attribute_id attribute) const {
```

Comment:
To be consistent with ``getEntryChecked``, you can name this function as ``getEntryUnchecked``